Return-Path: <SRS0=IoHm=TO=kvack.org=owner-linux-mm@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.6 (2021-04-09) on hp2
X-Spam-Level: 
X-Spam-Status: No, score=-102.0 required=2.5 tests=BAYES_00,
	FREEMAIL_FORGED_FROMDOMAIN,FREEMAIL_FROM,HEADER_FROM_DIFFERENT_DOMAINS,
	MAILING_LIST_MULTI,USER_IN_WELCOMELIST,USER_IN_WHITELIST autolearn=ham
	autolearn_force=no version=3.4.6
Received: from hp2 (localhost [127.0.0.1])
	by hp2 (8.17.1/8.17.1) with ESMTP id 223ML2pe163873
	for <akpm@localhost>; Thu, 3 Mar 2022 14:21:03 -0800
Received: from imap.fastmail.com [66.111.4.135]
	by hp2 with IMAP (fetchmail-6.4.26)
	for <akpm@localhost> (single-drop); Thu, 03 Mar 2022 14:21:03 -0800 (PST)
Received: from compute2.internal (compute2.nyi.internal [10.202.2.46])
	 by sloti43n14 (Cyrus 3.5.0-alpha0-4778-g14fba9972e-fm-20220217.001-g14fba997) with LMTPA;
	 Thu, 03 Mar 2022 17:20:24 -0500
X-Cyrus-Session-Id: sloti43n14-1646346024-897637-2-3883791085399549319
X-Sieve: CMU Sieve 3.0
X-Resolved-to: akpm@mbx.kernel.org
X-Delivered-to: akpm@mbx.kernel.org
X-Original-Delivered-to: linux-mm@kvack.org
X-Mail-from: SRS0=IoHm=TO=kvack.org=owner-linux-mm@kernel.org
Received: from mx2 ([10.202.2.201])
  by compute2.internal (LMTPProxy); Thu, 03 Mar 2022 17:20:24 -0500
Received: from mx2.messagingengine.com (localhost [127.0.0.1])
	by mailmx.nyi.internal (Postfix) with ESMTP id 5D5A61A2009B
	for <akpm@mbx.kernel.org>; Thu,  3 Mar 2022 17:20:24 -0500 (EST)
Received: from mx2.messagingengine.com (localhost [127.0.0.1])
    by mx2.messagingengine.com (Authentication Milter) with ESMTP
    id 803E9C6BD11;
    Thu, 3 Mar 2022 17:20:24 -0500
ARC-Seal: i=1; a=rsa-sha256; cv=none; d=messagingengine.com; s=fm2; t=
    1646346024; b=dnDsjezoJLyAcePjo/0ENHnq/LvsKT9mR9X+qegKAeQXM3kxDB
    OK9d1sj0Ezg81OKim4FdNv/RCa5m48EfS+QXl/5/2YUj/iH+mTjUvPttOIoZYkHO
    L4jEck6/sQDA+Jyx7exwwSaGJBfiVTZwN97CfjJvYOVjxVwtAEAUWnX7pyPf2Lxa
    KWTCt4ieUKhNksn7SCHrHHKqYogxmxDIMBc/EnPm9YApu5yrD6lX9cZbZz6M72KK
    jPClnjptLgQT1W8KyJSKk6eqZngua9x8UXxIk8SuGVNVZG9wLvg7GTg7eSRvwc+P
    yQm2ARVtaip2w/mp6OGi9CRe20+KC9JdYh7A==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=
    messagingengine.com; h=from:to:cc:subject:date:message-id
    :mime-version:content-transfer-encoding:sender:list-id; s=fm2;
    t=1646346024; bh=i+jhSHjqTSR2b7tP60Q6oAqpuTejuyJKZJEpX0NApkc=; b=
    AhcH+uu26MwmIJAay2G4T1fpVEJ/qCl9rrt6c1hyqwKDK8LVHGbOubjzTf6B7yEV
    tohdw4lU0OnEve0sCzgHNkCh4AVc9psisqouFQdUpe94BQrK/qLe0if0Q27RLAlv
    HOlBqj8gd26j9TAwlk1yiR+S8rSTPrniY86R4Z1laKRha2ywvZ1YkEYGtWW+b+SE
    VaLboL47AXxOkrFZ0T6FMmSIm37nQ8MmI2pjiVvCZjF20WQRZWjPM3e+PZYmFb1R
    exgpenMw0lbThjdn/7ke7oyQeUvj7pWx5JWZD00H4DUJ8kxzdh0BPp+EyZaAgBFS
    6XP2vjRk9ejEuqTu/ujTeA==
ARC-Authentication-Results: i=1; mx2.messagingengine.com;
    x-csa=none;
    x-me-sender=none;
    x-ptr=pass smtp.helo=dfw.source.kernel.org
    policy.ptr=dfw.source.kernel.org;
    bimi=skipped (DMARC Policy is not at enforcement);
    arc=none (no signatures found);
    dkim=pass (2048-bit rsa key sha256) header.d=gmail.com
    header.i=@gmail.com header.b=XyAJUTQ4 header.a=rsa-sha256
    header.s=20210112 x-bits=2048;
    dmarc=pass policy.published-domain-policy=none
    policy.published-subdomain-policy=quarantine
    policy.applied-disposition=none policy.evaluated-disposition=none
    (p=none,sp=quarantine,has-list-id=yes,d=none,d.eval=none)
    policy.policy-from=p header.from=gmail.com;
    iprev=pass smtp.remote-ip=139.178.84.217 (dfw.source.kernel.org);
    spf=pass
    smtp.mailfrom="SRS0=IoHm=TO=kvack.org=owner-linux-mm@kernel.org"
    smtp.helo=dfw.source.kernel.org
X-ME-Authentication-Results: mx2.messagingengine.com;
    x-aligned-from=fail;
    x-google-dkim=pass (2048-bit rsa key) header.d=1e100.net
      header.i=@1e100.net header.b=RGUWvtPN;
    x-return-mx=pass header.domain=gmail.com policy.is_org=yes
      (MX Records found: alt3.gmail-smtp-in.l.google.com,gmail-smtp-in.l.google.com,alt1.gmail-smtp-in.l.google.com,alt4.gmail-smtp-in.l.google.com,alt2.gmail-smtp-in.l.google.com);
    x-return-mx=pass smtp.domain=kernel.org policy.is_org=yes
      (MX Records found: smtp2.kernel.org,smtp3.kernel.org,smtp1.kernel.org);
    x-tls=pass smtp.version=TLSv1.2 smtp.cipher=ADH-AES256-GCM-SHA384
      smtp.bits=256/256;
    x-vs=clean score=0 state=0
Authentication-Results: mx2.messagingengine.com;
    x-csa=none;
    x-me-sender=none;
    x-ptr=pass smtp.helo=dfw.source.kernel.org
      policy.ptr=dfw.source.kernel.org
Authentication-Results: mx2.messagingengine.com;
    bimi=skipped (DMARC Policy is not at enforcement)
Authentication-Results: mx2.messagingengine.com;
    arc=none (no signatures found)
Authentication-Results: mx2.messagingengine.com;
    dkim=pass (2048-bit rsa key sha256) header.d=gmail.com
      header.i=@gmail.com header.b=XyAJUTQ4 header.a=rsa-sha256
      header.s=20210112 x-bits=2048;
    dmarc=pass policy.published-domain-policy=none
      policy.published-subdomain-policy=quarantine
      policy.applied-disposition=none policy.evaluated-disposition=none
      (p=none,sp=quarantine,has-list-id=yes,d=none,d.eval=none)
      policy.policy-from=p header.from=gmail.com;
    iprev=pass smtp.remote-ip=139.178.84.217 (dfw.source.kernel.org);
    spf=pass
      smtp.mailfrom="SRS0=IoHm=TO=kvack.org=owner-linux-mm@kernel.org"
      smtp.helo=dfw.source.kernel.org
X-ME-VSCause: gggruggvucftvghtrhhoucdtuddrgedvvddruddtiedgudeitdcutefuodetggdotefrod
    ftvfcurfhrohhfihhlvgemucfhrghsthforghilhdpggftfghnshhusghstghrihgsvgdp
    uffrtefokffrpgfnqfghnecuuegrihhlohhuthemuceftddtnecunecujfgurhephffvuf
    ffkffoggfgshhpjeestdekredtredttdenucfhrhhomhepjggrnhhgucfuhhhiuceoshhh
    hiekvdekfedtudesghhmrghilhdrtghomheqnecuggftrfgrthhtvghrnhepgfffhfelfe
    dtgeeuieelueevvefhveettdegudffgedufeeliedugfegtdetvdefnecuffhomhgrihhn
    pehkvghrnhgvlhdrohhrghenucfkphepudefledrudejkedrkeegrddvudejpdeijedrud
    ejgedrvdeguddrudegheenucevlhhushhtvghrufhiiigvpedvnecurfgrrhgrmhepihhn
    vghtpedufeelrddujeekrdekgedrvddujedphhgvlhhopegufhifrdhsohhurhgtvgdrkh
    gvrhhnvghlrdhorhhgpdhmrghilhhfrhhomhepoefutffutdepkfhojfhmpefvqfepkhhv
    rggtkhdrohhrghepohifnhgvrhdqlhhinhhugidqmhhmsehkvghrnhgvlhdrohhrgheq
X-ME-VSScore: 0
X-ME-VSCategory: clean
X-ME-CSA: none
Received-SPF: pass
    (kernel.org: Sender is authorized to use 'SRS0=IoHm=TO=kvack.org=owner-linux-mm@kernel.org' in 'mfrom' identity (mechanism 'include:_spf.kernel.org' matched))
    receiver=mx2.messagingengine.com;
    identity=mailfrom;
    envelope-from="SRS0=IoHm=TO=kvack.org=owner-linux-mm@kernel.org";
    helo=dfw.source.kernel.org;
    client-ip=139.178.84.217
Received: from dfw.source.kernel.org (dfw.source.kernel.org [139.178.84.217])
	(using TLSv1.2 with cipher ADH-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by mx2.messagingengine.com (Postfix) with ESMTPS
	for <akpm@mbx.kernel.org>; Thu,  3 Mar 2022 17:20:23 -0500 (EST)
Received: from smtp.kernel.org (relay.kernel.org [52.25.139.140])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by dfw.source.kernel.org (Postfix) with ESMTPS id 3337C60B20
	for <akpm@mbx.kernel.org>; Thu,  3 Mar 2022 22:20:22 +0000 (UTC)
Received: by smtp.kernel.org (Postfix)
	id A1F19C340EF; Thu,  3 Mar 2022 22:20:21 +0000 (UTC)
X-Remote-Delivered-To: akpm@kernel.org
Received: from kanga.kvack.org (kanga.kvack.org [205.233.56.17])
	by smtp.kernel.org (Postfix) with ESMTP id 8A050C340EC;
	Thu,  3 Mar 2022 22:20:20 +0000 (UTC)
DMARC-Filter: OpenDMARC Filter v1.4.1 smtp.kernel.org 8A050C340EC
Authentication-Results: smtp.kernel.org; dmarc=fail (p=none dis=none) header.from=gmail.com
Authentication-Results: smtp.kernel.org; spf=pass smtp.mailfrom=kvack.org
Received: by kanga.kvack.org (Postfix)
	id 8B4268D0002; Thu,  3 Mar 2022 17:20:19 -0500 (EST)
X-Remote-Delivered-To: linux-mm-outgoing@kvack.org
Received: by kanga.kvack.org (Postfix, from userid 40)
	id 863B18D0001; Thu,  3 Mar 2022 17:20:19 -0500 (EST)
X-Original-To: int-list-linux-mm@kvack.org
X-Remote-Delivered-To: int-list-linux-mm@kvack.org
Received: by kanga.kvack.org (Postfix, from userid 63042)
	id 72BE58D0002; Thu,  3 Mar 2022 17:20:19 -0500 (EST)
X-Original-To: linux-mm@kvack.org
X-Remote-Delivered-To: linux-mm@kvack.org
Received: from relay.hostedemail.com (relay.hostedemail.com [64.99.140.27])
	by kanga.kvack.org (Postfix) with ESMTP id 664A68D0001
	for <linux-mm@kvack.org>; Thu,  3 Mar 2022 17:20:19 -0500 (EST)
Received: from smtpin05.hostedemail.com (a10.router.float.18 [10.200.18.1])
	by unirelay10.hostedemail.com (Postfix) with ESMTP id 228FE20A5
	for <linux-mm@kvack.org>; Thu,  3 Mar 2022 22:20:19 +0000 (UTC)
X-FDA: 79204494558.05.FA14C72
Received: from mail-pl1-f181.google.com (mail-pl1-f181.google.com [209.85.214.181])
	by imf27.hostedemail.com (Postfix) with ESMTP id 981EB4000A
	for <linux-mm@kvack.org>; Thu,  3 Mar 2022 22:20:18 +0000 (UTC)
Received: by mail-pl1-f181.google.com with SMTP id 9so6039994pll.6
        for <linux-mm@kvack.org>; Thu, 03 Mar 2022 14:20:18 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20210112;
        h=from:to:cc:subject:date:message-id:mime-version
         :content-transfer-encoding;
        bh=i+jhSHjqTSR2b7tP60Q6oAqpuTejuyJKZJEpX0NApkc=;
        b=XyAJUTQ4WSbV0ZyMFvjiRuolDwAoBCGZtPOcZOv7DEcHXHhJ299eb8alyt7lrHlu59
         CbD3DKrch2N1D3gLJ/bGZE2r5qBFdEiXN7AFlrNeumYPduVhNF6RzWVDmrr1vUjuBJJX
         N3mhBWlPcKo/qIUnGACQl+mA3ssd1t/Dj/jODZNjNWLv6IWIS/N4ZNTIOCmY1LEHWG5F
         jB2s9TqRRwLl0ETtLcCEV8fM3IieaSAvH0gxWxn4ATX1kEcbuRwbi/UuA8LnpoSwLHoE
         O3z1/cByBoQlobWc0WOqoEhtSwwl6cTpNmkGxcVOaN5hBT2rpO63LFIHvjlu0bbXxdxj
         OuJQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20210112;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:mime-version
         :content-transfer-encoding;
        bh=i+jhSHjqTSR2b7tP60Q6oAqpuTejuyJKZJEpX0NApkc=;
        b=RGUWvtPNZkSAH6jCWAfr2sgcwPuZLs0nLwrPkvBnvBhMxEnyt4x8ENKMaf0uNUhr/l
         d6/orXQcLu2wfJ8BZGDrVe2llTWpy7xm9qr6kk2fVg3O2cKnAYuFMdyR4P6DPYLhVNiC
         77ulYWFslDAhasJ45fbQpKTsyEfd0IUTp6mACUFOLqFlSLa+a3lC8wP8BlQde7fhoWsP
         rpxDmLuAdjF2pZhiTvNVnMyVeq0aSn3J9eOGB1HZpzQdKMz4DkHLP5RLP1z0D6hU9guY
         9++prZKRelixTNZcAT5gwE48jUZvxRNwgeAKxO8ZUE1/N3ly2VPFWI+oi1gyVmoeT+C0
         oAJg==
X-Gm-Message-State: AOAM530Oqfl5uTPRKcgk/n9y1DWplNtR53I8WGiMfIS63GU0Z3THNPur
	kTbizRM8eUs3AQO4Jkgyo7g=
X-Google-Smtp-Source: ABdhPJwRrG9E7twGHq/irRPt/iu4Nnxg+Oyd48nX0497BwVHi+gDwkd77MOdUbrwvBTRn4gCGupj9A==
X-Received: by 2002:a17:903:2342:b0:14f:dec2:9849 with SMTP id c2-20020a170903234200b0014fdec29849mr28439484plh.74.1646346017514;
        Thu, 03 Mar 2022 14:20:17 -0800 (PST)
Received: from localhost.localdomain (c-67-174-241-145.hsd1.ca.comcast.net. [67.174.241.145])
        by smtp.gmail.com with ESMTPSA id lx7-20020a17090b4b0700b001b7d5b6d10asm2862172pjb.48.2022.03.03.14.20.15
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Thu, 03 Mar 2022 14:20:16 -0800 (PST)
From: Yang Shi <shy828301@gmail.com>
To: david@redhat.com, aarcange@redhat.com, hughd@google.com,
        kirill.shutemov@linux.intel.com, akpm@linux-foundation.org
Cc: shy828301@gmail.com, linux-mm@kvack.org, linux-kernel@vger.kernel.org
Subject: mm: thp: don't have to lock page anymore when splitting PMD
Date: Thu,  3 Mar 2022 14:20:14 -0800
Message-Id: <20220303222014.517033-1-shy828301@gmail.com>
X-Mailer: git-send-email 2.26.3
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Rspamd-Server: rspam10
X-Rspam-User: 
X-Stat-Signature: bitw8dmy6wgimmnkqbquhxfp11pon1q5
Authentication-Results: imf27.hostedemail.com;
	dkim=pass header.d=gmail.com header.s=20210112 header.b=XyAJUTQ4;
	spf=pass (imf27.hostedemail.com: domain of shy828301@gmail.com designates 209.85.214.181 as permitted sender) smtp.mailfrom=shy828301@gmail.com;
	dmarc=pass (policy=none) header.from=gmail.com
X-Rspamd-Queue-Id: 981EB4000A
X-HE-Tag: 1646346018-363788
X-Bogosity: Ham, tests=bogofilter, spamicity=0.000000, version=1.2.4
Sender: owner-linux-mm@kvack.org
Precedence: bulk
X-Loop: owner-majordomo@kvack.org
List-ID: <linux-mm.kvack.org>

The commit c444eb564fb1 ("mm: thp: make the THP mapcount atomic against
__split_huge_pmd_locked()") locked the page for PMD split to make
mapcount stable for reuse_swap_page(), then commit 1c2f67308af4 ("mm:
thp: fix MADV_REMOVE deadlock on shmem THP") reduce the scope to
anonymous page only.

However COW has not used mapcount to determine if the page is shared or
not anymore due to the COW fixes [1] from David Hildenbrand and the
reuse_swap_page() was removed as well.  So PMD split doesn't have to
lock the page anymore.  This patch basically reverted the above two
commits.

[1] https://lore.kernel.org/linux-mm/20220131162940.210846-1-david@redhat.com/

Cc: David Hildenbrand <david@redhat.com>
Cc: Andrea Arcangeli <aarcange@redhat.com>
Cc: Hugh Dickins <hughd@google.com>
Cc: "Kirill A . Shutemov" <kirill.shutemov@linux.intel.com>
Signed-off-by: Yang Shi <shy828301@gmail.com>
---
 mm/huge_memory.c | 44 +++++---------------------------------------
 1 file changed, 5 insertions(+), 39 deletions(-)

diff --git a/mm/huge_memory.c b/mm/huge_memory.c
index b49e1a11df2e..daaa698bd273 100644
--- a/mm/huge_memory.c
+++ b/mm/huge_memory.c
@@ -2134,8 +2134,6 @@ void __split_huge_pmd(struct vm_area_struct *vma, pmd_t *pmd,
 {
 	spinlock_t *ptl;
 	struct mmu_notifier_range range;
-	bool do_unlock_folio = false;
-	pmd_t _pmd;
 
 	mmu_notifier_range_init(&range, MMU_NOTIFY_CLEAR, 0, vma, vma->vm_mm,
 				address & HPAGE_PMD_MASK,
@@ -2148,48 +2146,16 @@ void __split_huge_pmd(struct vm_area_struct *vma, pmd_t *pmd,
 	 * pmd against. Otherwise we can end up replacing wrong folio.
 	 */
 	VM_BUG_ON(freeze && !folio);
-	if (folio) {
-		VM_WARN_ON_ONCE(!folio_test_locked(folio));
-		if (folio != page_folio(pmd_page(*pmd)))
-			goto out;
-	}
+	if (folio && folio != page_folio(pmd_page(*pmd)))
+		goto out;
 
-repeat:
-	if (pmd_trans_huge(*pmd)) {
-		if (!folio) {
-			folio = page_folio(pmd_page(*pmd));
-			/*
-			 * An anonymous page must be locked, to ensure that a
-			 * concurrent reuse_swap_page() sees stable mapcount;
-			 * but reuse_swap_page() is not used on shmem or file,
-			 * and page lock must not be taken when zap_pmd_range()
-			 * calls __split_huge_pmd() while i_mmap_lock is held.
-			 */
-			if (folio_test_anon(folio)) {
-				if (unlikely(!folio_trylock(folio))) {
-					folio_get(folio);
-					_pmd = *pmd;
-					spin_unlock(ptl);
-					folio_lock(folio);
-					spin_lock(ptl);
-					if (unlikely(!pmd_same(*pmd, _pmd))) {
-						folio_unlock(folio);
-						folio_put(folio);
-						folio = NULL;
-						goto repeat;
-					}
-					folio_put(folio);
-				}
-				do_unlock_folio = true;
-			}
-		}
-	} else if (!(pmd_devmap(*pmd) || is_pmd_migration_entry(*pmd)))
+	if (!(pmd_devmap(*pmd) || is_pmd_migration_entry(*pmd)))
 		goto out;
+
 	__split_huge_pmd_locked(vma, pmd, range.start, freeze);
 out:
 	spin_unlock(ptl);
-	if (do_unlock_folio)
-		folio_unlock(folio);
+
 	/*
 	 * No need to double call mmu_notifier->invalidate_range() callback.
 	 * They are 3 cases to consider inside __split_huge_pmd_locked():
-- 
2.26.3

